### PR TITLE
Add support for Realm.executeTransaction using coroutines and RealmResults.toFlow

### PR DESCRIPTION
**Pull request by edualonso**
Added draft versions of:

- `Realm.executeTransactionAwait` - suspend extension function that uses the existing Java SDK for running transactions.
- `RealmResults.toFlow` - coroutine-equivalent to `asFlowable`.

Added also a coroutine example to the examples project.

### CODE REVIEW

File: realm/kotlin-extensions/build.gradle [hunk 1]
```diff
@@ -26,6 +26,7 @@ android {
         minSdkVersion rootProject.minSdkVersion
         targetSdkVersion rootProject.compileSdkVersion
         versionName version
+        multiDexEnabled true
```
[the remark this answers is not in the document]
Comment (cmelchior) [reply realm/kotlin-extensions/build.gradle:29]: Also, how does Multidex work on libraries if you include it in an app that doesn't define Multidex?

File: realm/kotlin-extensions/src/androidTest/kotlin/io/realm/CoroutineTests.kt [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+
+    @After
+    fun tearDown() {
+        realm.executeTransaction { it.deleteAll() }
+        realm.close()
+    }
+
+    @Test
+    fun toFlow_emittedOnCollect() {
+        Realm.getDefaultInstance().use { defaultRealm ->
```
Comment (cmelchior) [inline realm/kotlin-extensions/src/androidTest/kotlin/io/realm/CoroutineTests.kt:60]: We should not use `getDefaultInstance()` but the Realm setup in the `setUp` method. Same in all the other tests.

File: realm/kotlin-extensions/src/androidTest/kotlin/io/realm/CoroutineTests.kt [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+                    .onEach { flowResults ->
+                        assertTrue(flowResults.isFrozen)
+                        if (flowResults.size == 2) {
+                            scope.cancel()
+                        }
+                    }.onCompletion {
+                        countDownLatch.countDown()
+                    }.launchIn(scope)
+        }
+        countDownLatch.await()
```
Comment (cmelchior) [inline realm/kotlin-extensions/src/androidTest/kotlin/io/realm/CoroutineTests.kt:85]: use `TestHelper.awaitOrThrow(countDownLatch)` to avoid blocking CI for 1.5 hours if something in this test doesn't work.

File: realm/kotlin-extensions/src/androidTest/kotlin/io/realm/CoroutineTests.kt [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+            }
+        }
+
+        val countDownLatch = CountDownLatch(1)
+
+        // TODO check this out for better testing: https://proandroiddev.com/from-rxjava-to-kotlin-flow-testing-42f1641d8433
+        val scope = CoroutineScope(Dispatchers.Main)
+        scope.launch {
+            Realm.getDefaultInstance()
+                    .where(SimpleClass::class.java)
```
Comment (cmelchior) [inline realm/kotlin-extensions/src/androidTest/kotlin/io/realm/CoroutineTests.kt:73]: Use `realm.where<SimpleClass>()`, this also applies to a lot of the other tests

File: realm/kotlin-extensions/src/androidTest/kotlin/io/realm/CoroutineTests.kt [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+
+        // Get results from the test thread
+        val findAll = Realm.getDefaultInstance()
+                .where(SimpleClass::class.java)
+                .findAll()
+
+        CoroutineScope(Dispatchers.Main).launch {
+            assertFailsWith<IllegalStateException> {
+                // Now we are on the main thread, which means crash
+                findAll.toFlow().collect()
```
Comment (cmelchior) [inline realm/kotlin-extensions/src/androidTest/kotlin/io/realm/CoroutineTests.kt:100]: Don't also call `collect()` if the intend is that it is `toFlow()` that throws.

File: realm/kotlin-extensions/src/androidTest/kotlin/io/realm/CoroutineTests.kt [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+            }
+            countDownLatch.countDown()
+        }
+        countDownLatch.await()
+    }
+
+    @Test
+    fun executeTransactionAwait() {
+        testScope.runBlockingTest {
+            Realm.getDefaultInstance().use { defaultRealm ->
```
Comment (cmelchior) [inline realm/kotlin-extensions/src/androidTest/kotlin/io/realm/CoroutineTests.kt:111]: Use `Realm.getInstance(realm.configuration)` if you need to use a configuration from the existing Realm or save the configuration in an test class property so you can do `Realm.getInstance(config)`

File: realm/kotlin-extensions/src/androidTest/kotlin/io/realm/CoroutineTests.kt [hunk 7]
```diff
@@ -0,0 +1,147 @@
...
+                assertEquals(0, defaultRealm.where(SimpleClass::class.java).findAll().size)
+
+                defaultRealm.executeTransactionAwait(testDispatcher) { transactionRealm ->
+                    val simpleObject = SimpleClass().apply { name = "simpleName" }
+                    transactionRealm.insert(simpleObject)
+                }
+            }
+        }
+        assertEquals(1, Realm.getDefaultInstance().where(SimpleClass::class.java).findAll().size)
```
Comment (cmelchior) [inline realm/kotlin-extensions/src/androidTest/kotlin/io/realm/CoroutineTests.kt:120]: IMO this assert should be right after you called `defaultRealm.executeTransactionAwait` to make sure that the result is available immediately.

File: realm/kotlin-extensions/src/androidTest/kotlin/io/realm/CoroutineTests.kt [hunk 8]
```diff
@@ -0,0 +1,147 @@
...
+        // It will crash so long we aren't using Dispatchers.Unconfined
+        CoroutineScope(Dispatchers.IO).launch {
+            assertFailsWith<IllegalStateException> {
+                defaultRealm.where(SimpleClass::class.java).findAll()
+                countDownLatch.countDown()
+            }.let {
+                exception = it
+            }
+        }
+        countDownLatch.await()
```
Comment (cmelchior) [inline realm/kotlin-extensions/src/androidTest/kotlin/io/realm/CoroutineTests.kt:142]: Use `TestHelper.awaitOrThrow`

File: realm/kotlin-extensions/src/androidTest/kotlin/io/realm/KotlinRealmTests.kt [hunk 9]
```diff
@@ -5,31 +5,42 @@ import io.realm.entities.PrimaryKeyClass
...
     fun setUp() {
         realm = Realm.getInstance(configFactory.createConfiguration())
     }
 
     @After
     fun tearDown() {
+        realm.executeTransaction { it.deleteAll() }
         realm.close()
+
+        with(Realm.getDefaultInstance()) {
```
Comment (cmelchior) [inline realm/kotlin-extensions/src/androidTest/kotlin/io/realm/KotlinRealmTests.kt:40]: Don't use `getDefaultInstance()`

File: realm/kotlin-extensions/src/androidTest/kotlin/io/realm/rule/BlockingLooperThread.kt [hunk 10]
```diff
@@ -0,0 +1,366 @@
...
+ *     // Runs after test completed or failed
+ * }
+ *
+ * @Test
+ * fun myTest() = looperThread.run {
+ *     // test code
+ * }
+ * ```
+ */
+class BlockingLooperThread {
```
Comment (cmelchior) [inline realm/kotlin-extensions/src/androidTest/kotlin/io/realm/rule/BlockingLooperThread.kt:59]: We should avoid copying files between `realm-library` and `kotlin-extensions`. In the past we used a shared sourceset to keep these files in. Doesn't that work?

File: realm/kotlin-extensions/src/main/kotlin/io/realm/kotlin/RealmExtensions.kt [hunk 11]
```diff
@@ -91,15 +100,37 @@ inline fun <reified T : RealmModel> Realm.createEmbeddedObject(parentObject: Rea
     return this.createEmbeddedObject(T::class.java, parentObject, parentProperty)
 }
 
+/**
+ * FIXME
```
Comment (cmelchior) [inline realm/kotlin-extensions/src/main/kotlin/io/realm/kotlin/RealmExtensions.kt:104]: Missing?

File: realm/kotlin-extensions/src/main/kotlin/io/realm/kotlin/RealmExtensions.kt [hunk 12]
```diff
@@ -91,15 +100,37 @@ inline fun <reified T : RealmModel> Realm.createEmbeddedObject(parentObject: Rea
...
 /**
 TODO: Figure out if we should include this is or not. Using this makes it possible to do
 
 inline fun <T> Realm.callTransaction(crossinline action: Realm.() -> T): T {
-    val ref = AtomicReference<T>()
-    executeTransaction {
-        ref.set(action(it))
-    }
-    return ref.get()
+val ref = AtomicReference<T>()
```
Comment (cmelchior) [inline realm/kotlin-extensions/src/main/kotlin/io/realm/kotlin/RealmExtensions.kt:129]: Delete this?

File: realm/kotlin-extensions/src/main/kotlin/io/realm/kotlin/RealmResultsExtensions.kt [hunk 13]
```diff
@@ -0,0 +1,103 @@
...
+}
+
+/*
+ *
+ * ALTERNATIVE IMPLEMENTATION USING flow INSTEAD OF callbackFlow
+ *
+ *
+
+@Beta
+fun <T : RealmModel> RealmResults<T>.toFlowNew(): Flow<RealmResults<T>> {
```
Comment (cmelchior) [inline realm/kotlin-extensions/src/main/kotlin/io/realm/kotlin/RealmResultsExtensions.kt:71]: Remove this?

File: realm/realm-library/src/main/java/io/realm/BaseRealm.java [hunk 14]
```diff
@@ -73,8 +73,9 @@
 
     static volatile Context applicationContext;
 
+    // FIXME: document this
     // Thread pool for all async operations (Query & transaction)
-    static final RealmThreadPoolExecutor asyncTaskExecutor = RealmThreadPoolExecutor.newDefaultExecutor();
+    public static final RealmThreadPoolExecutor asyncTaskExecutor = RealmThreadPoolExecutor.newDefaultExecutor();
```
Comment (cmelchior) [inline realm/realm-library/src/main/java/io/realm/BaseRealm.java:78]: Should we have two thread pools? One for writes and one for Reads? The pool for writes just need 1 thread since you cannot parallelize these anyway.

**Review by cmelchior — CHANGES_REQUESTED** [review]
Nice 👍 

A few comments:
- There are a few common patterns in the tests that need to change: use of default Realm, not using `TestHelper.awaitOrThrow` and not using `Realm.where<ModelClass>()`
- The tests, in general, look like they cover the things we want, but some tests around e.g. canceling a coroutine would be nice as well, e.g. in a loop that ensure that we hopefully hit the edge cases
- There seem to be missing Flow support on Lists and RealmObjects. They should just follow the same pattern as RealmResults.

File: examples/coroutinesExample/src/main/java/io/realm/examples/coroutinesexample/repository/Repository.kt [hunk 15]
```diff
@@ -0,0 +1,48 @@
...
+    override fun getDogs(): Flow<List<Doggo>> {
+        return Realm.getDefaultInstance()
+                .where(Doggo::class.java)
+                .findAllAsync()
+                .toFlow()
+    }
+
+    override fun deleteDogs() {
+        Realm.getDefaultInstance().use { defaultRealm ->
+            defaultRealm.executeTransaction { realm ->
```
[the remark this answers is not in the document]
Comment (edualonso) [reply examples/coroutinesExample/src/main/java/io/realm/examples/coroutinesexample/repository/Repository.kt:43]: That is right, I completely missed it.

**Review by rorbech — COMMENTED** [review]
Nice. As already pointed out, the tests seems a bit sparse on cancellation. Can't really remember the semantics of flows, but maybe also worth to consider if there is a need for cases around multiple consumers and re-registering to existing flows.

File: realm/kotlin-extensions/build.gradle (hunk 1 above)
[the remark this answers is not in the document]
Comment (edualonso) [reply realm/kotlin-extensions/build.gradle:29]: The extensions project won't compile unless added. Interestingly, the gradle dependency isn't needed though.

Also, I'm not sure this setting can be set only for tests.

[the remark this answers is not in the document]
Comment (edualonso) [reply realm/kotlin-extensions/build.gradle:29]: Not sure about how a non-multidex app behaves with a multidex library, I need to test it.

File: realm/realm-library/src/main/java/io/realm/BaseRealm.java (hunk 14 above)
Comment (edualonso) [reply realm/realm-library/src/main/java/io/realm/BaseRealm.java]: I've ended up adding one public thread pool with one thread which is used only for `executeTransactionAwait`. I don't think it's needed to add an additional one for reads as flows need to be collected on the same thread the query was performed.

**Comment by edualonso** [comment]
Ready for a second round, please give it a go folks.

File: realm/kotlin-extensions/src/main/kotlin/io/realm/kotlin/RealmObjectExtensions.kt [hunk 16]
```diff
@@ -0,0 +1,151 @@
...
+        val proxy = obj as RealmObjectProxy
+        val realm = proxy.`realmGet$proxyState`().`realm$realm`
+
+        when (realm) {
+            is Realm -> flowFromRealm<T>(realm, obj)
+            is DynamicRealm -> {
+                val dynamicObject = obj as DynamicRealmObject
+                flowFromDynamicRealm(realm, dynamicObject) as Flow<T>
+            }
+            // FIXME: exception text
```
Comment (cmelchior) [inline realm/kotlin-extensions/src/main/kotlin/io/realm/kotlin/RealmObjectExtensions.kt:71]: We don't have tests for this elsewhere. Just deleting this comment would be good enough.

File: realm/kotlin-extensions/src/main/kotlin/io/realm/kotlin/RealmObjectExtensions.kt [hunk 17]
```diff
@@ -0,0 +1,151 @@
...
+            else -> throw UnsupportedOperationException("${realm.javaClass} does not support RxJava. See https://realm.io/docs/java/latest/#rxjava for more details.")
+        }
+    } else {
+        // TODO Is this true? Should we just return Observable.just(object) ?
```
Comment (cmelchior) [inline realm/kotlin-extensions/src/main/kotlin/io/realm/kotlin/RealmObjectExtensions.kt:75]: Good point. I can see that we disallow this for RxJava Observables as well. I cannot remember exactly why we did that.

This seems similar to frozen objects so probably fine to allow?

File: realm/realm-library/src/main/java/io/realm/BaseRealm.java [hunk 18]
```diff
@@ -76,6 +76,12 @@
     // Thread pool for all async operations (Query & transaction)
     static final RealmThreadPoolExecutor asyncTaskExecutor = RealmThreadPoolExecutor.newDefaultExecutor();
 
+    /**
+     * Thread pool executor used for write operations - only one thread is needed as writes cannot
+     * be parallelized.
+     */
+    public static final RealmThreadPoolExecutor writeExecutor = RealmThreadPoolExecutor.newSingleThreadExecutor();
```
Comment (cmelchior) [inline realm/realm-library/src/main/java/io/realm/BaseRealm.java:83]: We should use `WRITE_EXECUTOR` since this is public. If we introduce this, we should probably also redirect `executeTransactionAsync` to use it.

Comment (cmelchior) [reply realm/realm-library/src/main/java/io/realm/BaseRealm.java:83]: But it is fine to create an issue for that and do it in a followup PR

**Review by cmelchior — APPROVED** [review]
Woop woop. Nice 🎉  Had a few minor comments.

File: realm/kotlin-extensions/src/androidTest/kotlin/io/realm/CoroutineTests.kt [hunk 19]
```diff
@@ -0,0 +1,510 @@
...
+                    .toFlow()
+                    .flowOn(context)
+                    .onEach { flowResults ->
+                        assertTrue(flowResults.isFrozen)
+                        assertEquals(0, flowResults.size)
+                        scope.cancel("Cancelling scope...")
+                    }.onCompletion {
+                        realmInstance.close()
+                        countDownLatch.countDown()
+                    }.launchIn(scope)
```
[the remark this answers is not in the document]
Comment (edualonso) [reply realm/kotlin-extensions/src/androidTest/kotlin/io/realm/CoroutineTests.kt:69]: Maybe this one is not the best way to start collecting results from a flow, but `launchIn` is shorthand for `scope.launch { flow.collect() }`. The advantage is that it returns a `Job`, so that may come in handy for other purposes. But yeah, in this case I'm not doing anything with the job, so I might as well use collect instead.

[the remark this answers is not in the document]
Comment (edualonso) [reply realm/kotlin-extensions/src/androidTest/kotlin/io/realm/CoroutineTests.kt:69]: There is a difference though, which is `launchIn` is not a coroutine and `collect` is.

File: realm/kotlin-extensions/src/androidTest/kotlin/io/realm/CoroutineTests.kt [hunk 20]
```diff
@@ -0,0 +1,510 @@
...
+            val realmInstance = Realm.getInstance(configuration)
+
+            realmInstance.where<SimpleClass>()
+                    .findAllAsync()
+                    .toFlow()
+                    .flowOn(context)
+                    .onEach { flowResults ->
+                        assertTrue(flowResults.isFrozen)
+                        assertEquals(0, flowResults.size)
+                        scope.cancel("Cancelling scope...")
```
[the remark this answers is not in the document]
Comment (edualonso) [reply realm/kotlin-extensions/src/androidTest/kotlin/io/realm/CoroutineTests.kt:65]: Yep, calling `cancel` on a scope will trigger `onComplete`, otherwise the flow will never end.

File: realm/kotlin-extensions/src/androidTest/kotlin/io/realm/CoroutineTests.kt [hunk 21]
```diff
@@ -0,0 +1,510 @@
...
+        val context = Dispatchers.Main
+        val scope = CoroutineScope(context)
+
+        scope.launch {
+            val realmInstance = Realm.getInstance(configuration)
+            realmInstance.where<SimpleClass>()
+                    .findAllAsync()
+                    .toFlow()
+                    .flowOn(context)
+                    .onCompletion {
```
[the remark this answers is not in the document]
Comment (edualonso) [reply realm/kotlin-extensions/src/androidTest/kotlin/io/realm/CoroutineTests.kt:120]: `toFlow` emits whatever results we have immediately, and normally it is an empty result the first time, which means `onEach` will be called most likely at least once with empty `RealmResults`. However, since this test is tightly synchronised - look at the `delay(100)` just a few lines below - maybe it doesn't even make sense to test it...?

File: realm/kotlin-extensions/src/androidTest/kotlin/io/realm/CoroutineTests.kt (hunk 19 above)
[the remark this answers is not in the document]
Comment (edualonso) [reply realm/kotlin-extensions/src/androidTest/kotlin/io/realm/CoroutineTests.kt:69]: This actually means that if we call `collect` and the flow isn't cancelled somehow, it will never end and will hog the thread on which it's running, which isn't a problem here, but it is for instance in `toFlow_resultsCancelBeforeCollectActualResults` since we are doing things within the coroutine scope after calling collect.

**Review by cmelchior — APPROVED** [review]
